Bottom: 46a41435ac366ad36c28dcab41598bb149a98d57
Top:    7c61c1c99e415d20eed171e106378d1cece5b80c
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-05-21 19:02:13 -0700

Refresh of pre-qpsize

---

diff --git a/src/preload.c b/src/preload.c
index 87778eb..12f2c02 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -97,6 +97,9 @@ static int (*real_fcntl)(int socket, int cmd, ... /* arg */);
 static struct index_map idm;
 static pthread_mutex_t mut = PTHREAD_MUTEX_INITIALIZER;
 
+static int sq_size;
+static int rq_size;
+
 enum fd_type {
 	fd_normal,
 	fd_rsocket
@@ -196,6 +199,19 @@ static enum fd_type fd_close(int index, int *fd)
 	return type;
 }
 
+void getenv_options(void)
+{
+	char *var;
+
+	var = getenv("RS_SQ_SIZE");
+	if (var)
+		sq_size = atoi(var);
+
+	var = getenv("RS_RQ_SIZE");
+	if (var)
+		rq_size = atoi(var);
+}
+
 static void init_preload(void)
 {
 	static int init;
@@ -231,6 +247,8 @@ static void init_preload(void)
 	real_setsockopt = dlsym(RTLD_NEXT, "setsockopt");
 	real_getsockopt = dlsym(RTLD_NEXT, "getsockopt");
 	real_fcntl = dlsym(RTLD_NEXT, "fcntl");
+
+	getenv_options();
 	init = 1;
 out:
 	pthread_mutex_unlock(&mut);
@@ -285,6 +303,24 @@ err:
 	return ret;
 }
 
+/*
+ * Use defaults on failure.
+ */
+void set_rsocket_options(int rsocket)
+{
+	socklen_t len;
+
+	if (sq_size) {
+		len = sizeof(sq_size);
+		rsetsockopt(rsocket, SOL_RDMA, RDMA_SQSIZE, &sq_size, &len);
+	}
+
+	if (rq_size) {
+		len = sizeof(rq_size);
+		rsetsockopt(rsocket, SOL_RDMA, RDMA_RQSIZE, &rq_size, &len);
+	}
+}
+
 int socket(int domain, int type, int protocol)
 {
 	int index, ret;
@@ -297,6 +333,7 @@ int socket(int domain, int type, int protocol)
 	ret = rsocket(domain, type, protocol);
 	if (ret >= 0) {
 		fd_store(index, ret, fd_rsocket);
+		set_rsocket_options(ret);
 		return index;
 	} else {
 		fd_close(index, &ret);
